### DISCUSSION TITLE
Question, not bug: Hit tab once to insert to end of common prefix, hit tab again to start cycling through options -- How to implement? Can tab perform this double duty?

### DISCUSSION THREAD

**Discussion by cohml** [marlonrichert/zsh-autocomplete#633] [Q&A] [answered]
Sorry, this is neither a bug nor a feature request. Rather, this is a behavior I believe is possible, but which I can't figure out how to configure, despite scouring documentation and issues. Not sure where/how else to ask.

Anyway, imagine a directory like this:

```
❯ ls -1
foo_a.py
foo_a.sh
foo_b.py
foo_b.sh
```

If I enter `file f` and hit tab, it should autocomplete to the longest prefix shared by all files starting with `f`:

```
❯ file foo_
```

Then, with the longest common prefix already completed, subsequent tab presses should cycle through the files:

```
❯ file foo_a.py
(hit tab)
❯ file foo_b.py
(hit tab)
❯ file foo_a.sh
(etc.)
```

Alternatively, imagine after getting `file foo_` I entered `file foo_a.`. Then hitting tab would cycle only through the matching files, because no more prefix remains to match: 

```
❯ file foo_a.py
(hit tab)
❯ file foo_a.sh
(hit tab)
❯ file foo_a.py
(etc.)
```

Can you advise on the keybindings or any other settings I should configure to bring about this behavior?

Note, if it matters: I use `bindkey -v`, so `viins` vs. `vicmd` adds an extra layer of complication.

**Comment by marlonrichert** [comment]
This is all documented in the Readme:
* [First insert the common substring](https://github.com/marlonrichert/zsh-autocomplete#first-insert-the-common-substring)
* [Make Tab and ShiftTab cycle completions on the command line](https://github.com/marlonrichert/zsh-autocomplete#make-tab-and-shifttab-cycle-completions-on-the-command-line)

**Reply by cohml** [reply]
> [First insert the common substring](https://github.com/marlonrichert/zsh-autocomplete#first-insert-the-common-substring)

I saw this, but when I add those lines to my RC file though (after sourcing `zsh-autocomplete`), it inserts the entire "common substring", not just the common prefix.

```
❯ ls -1
foo_a.sh
foo_b.sh
❯ ls f <TAB>
❯ ls foo_.sh
```

The desired behavior upon the first `<TAB>` is that it would insert only the common prefix `foo_`, not the common prefix `foo_` AND the common suffix `.sh`. There is no file named `foo_.sh`, so the common substring isn't really useful.

Sorry for the confusion.

Note that what I'm describing is just zsh's default tab completion behavior, I believe. I'd essentially like to retain that default behavior, but also have `zsh-autocomplete` show me the possible completions underneath. On my reading, the README does not cover this.

**Reply by marlonrichert** [reply]
That’s a result of using more complex completion matchers. The only thing that Zsh let’s you insert is the longest string that completes to all of generated matches. Znap has no control over that.

If you want to get a prefix only, you need modify the following setting to not complete any characters between the ones you’ve already typed:
https://github.com/marlonrichert/zsh-autocomplete/blob/453df9ed248534d51d4efc11864ce9d4d5192f6c/Functions/Init/.autocomplete%3Aconfig#L61

For documentation, see the Zsh manual:
* [Completion Widgets > Completion Matching Control](https://zsh.sourceforge.io/Doc/Release/Completion-Widgets.html#Completion-Matching-Control)
* [Completion System > Standard Styles](https://zsh.sourceforge.io/Doc/Release/Completion-System.html#Standard-Styles)

**Reply by s624** [reply]
Hey was looking for the answer to the same same question. The expected behavior would be pressing TAB for the second time would put the cursor where the match ended without the suffix.  Pressing TAB for second time does put the cursor there for a very short glimpse, but then the cursor comes back to the end position. 
@marlonrichert  Can you give us some direction how can we get the expected behavior, please? @mcauley-penney  was also looking for the same thing, I see.  

thank you. By the way your plugin is very very helpful. :)

**Reply by marlonrichert** [reply, marked as answer]
I've now added [instructions for this to the Readme](https://github.com/marlonrichert/zsh-autocomplete/commit/c7b65508fd3a016dc9cdb410af9ee7806b3f9be1)

**Reply by mcauley-penney** [reply]
This is working for me

**Reply by cohml** [reply]
And for me as well. Thanks @marlonrichert!

**Comment by mcauley-penney** [comment]
I'm also looking for this. You said

> That’s a result of using more complex completion matchers

But that seems to be untrue. With the default settings, I also get this behavior. I would appreciate if you could point me towards the complex completion matchers that are providing this. 

If this is unintentional, would it be possible to move it to bug status?

**Reply by marlonrichert** [reply]
> With the default settings, I also get this behavior.

Default settings of what? Autocomplete or Zsh?
